pkg/cgroups/manager/manager.go

```diff
@@ -92,6 +99,7 @@ func (m *CgroupManager) OnAddPod(pod *v1.Pod) {
 	if pod.Spec.NodeName != nodetypes.GetName() {
 		return
 	}
+	m.wg.Add(1)
```

Comment:
The use of WaitGroup here seems a bit suspect, specifically via [sync package docs](https://pkg.go.dev/sync#WaitGroup.Add):

>  If a WaitGroup is reused to wait for several independent sets of events, new Add calls must happen after all previous Wait calls have returned. See the WaitGroup example.

I'm not sure we can guarantee this if we have various different Add/Update/Delete calls coming from potentially different goroutines(?) without synchronizing with the waiter on the group.

The canonical usage of WaitGroups is to avoid storing them long-term, and instead create them on demand where the waiting is required, synchronously increase the wait count, then hand the WaitGroup to other goroutines to complete their piece via `Done()`: https://gobyexample.com/waitgroups .

If using WaitGroups, I'd suggest adding it as a parameter to these functions and copying it into to the shorter-lived `podEvent` structure. Or to match the `GetPodMetadataForContainer()` / `DumpPodMetadata()` style, you could create channels on-the-fly in a similar manner and return them. For example (incremental, only compile/unit tested):

```
diff --git a/pkg/cgroups/manager/manager.go b/pkg/cgroups/manager/manager.go
index 045a7e0ca66a..01234cd6a2d5 100644
--- a/pkg/cgroups/manager/manager.go
+++ b/pkg/cgroups/manager/manager.go
@@ -52,8 +52,6 @@ type CgroupManager struct {
        containerMetadataByCgrpId map[uint64]*containerMetadata
        // Buffered channel to receive pod events
        podEvents chan podEvent
-       // Tracks pod events being processed. Mainly used for testing.
-       wg sync.WaitGroup
        // Cgroup path provider
        pathProvider cgroupPathProvider
        // Object to get cgroup path provider
@@ -96,37 +94,55 @@ func NewCgroupManager() *CgroupManager {
 }
 
 func (m *CgroupManager) OnAddPod(pod *v1.Pod) {
+       m.onAddPod(pod)
+}
+
+func (m *CgroupManager) onAddPod(pod *v1.Pod) (done chan struct{}) {
        if pod.Spec.NodeName != nodetypes.GetName() {
                return
        }
-       m.wg.Add(1)
+       doneChan := make(chan struct{})
        m.podEvents <- podEvent{
                pod:       pod,
                eventType: podAddEvent,
+               done:      doneChan,
        }
+       return doneChan
 }
 
 func (m *CgroupManager) OnUpdatePod(oldPod, newPod *v1.Pod) {
+       m.onUpdatePod(oldPod, newPod)
+}
+
+func (m *CgroupManager) onUpdatePod(oldPod, newPod *v1.Pod) (done chan struct{}) {
        if newPod.Spec.NodeName != nodetypes.GetName() {
                return
        }
-       m.wg.Add(1)
+       doneChan := make(chan struct{})
        m.podEvents <- podEvent{
                pod:       newPod,
                oldPod:    oldPod,
                eventType: podUpdateEvent,
+               done:      doneChan,
        }
+       return doneChan
 }
 
 func (m *CgroupManager) OnDeletePod(pod *v1.Pod) {
+       m.onDeletePod(pod)
+}
+
+func (m *CgroupManager) onDeletePod(pod *v1.Pod) (done chan struct{}) {
        if pod.Spec.NodeName != nodetypes.GetName() {
                return
        }
-       m.wg.Add(1)
+       doneChan := make(chan struct{})
        m.podEvents <- podEvent{
                pod:       pod,
                eventType: podDeleteEvent,
+               done:      doneChan,
        }
+       return doneChan
 }
 
 // GetPodMetadataForContainer returns pod metadata for the given container
@@ -144,7 +160,6 @@ func (m *CgroupManager) GetPodMetadataForContainer(cgroupId uint64) *PodMetadata
 
        podMetaOut := make(chan *PodMetadata)
 
-       m.wg.Add(1)
        m.podEvents <- podEvent{
                cgroupId:       cgroupId,
                eventType:      podGetMetadataEvent,
@@ -160,7 +175,6 @@ func (m *CgroupManager) DumpPodMetadata() []*FullPodMetadata {
        }
        allMetaOut := make(chan []*FullPodMetadata)
 
-       m.wg.Add(1)
        m.podEvents <- podEvent{
                eventType:      podDumpMetadataEvent,
                allMetadataOut: allMetaOut,
@@ -195,6 +209,7 @@ type podEvent struct {
        eventType      int
        podMetadataOut chan *PodMetadata
        allMetadataOut chan []*FullPodMetadata
+       done           chan struct{}
 }
 
 type fs interface {
@@ -257,7 +272,9 @@ func (m *CgroupManager) processPodEvents() {
                select {
                case ev := <-m.podEvents:
                        if !m.enabled {
-                               m.wg.Done()
+                               if ev.done != nil {
+                                       close(ev.done)
+                               }
                                continue
                        }
                        switch ev.eventType {
@@ -270,7 +287,9 @@ func (m *CgroupManager) processPodEvents() {
                        case podDumpMetadataEvent:
                                m.dumpPodMetadata(ev.allMetadataOut)
                        }
-                       m.wg.Done()
+                       if ev.done != nil {
+                               close(ev.done)
+                       }
                case <-m.shutdown:
                        return
                }
diff --git a/pkg/cgroups/manager/manager_test.go b/pkg/cgroups/manager/manager_test.go
index c71ebcda36bd..b1d9a327ac46 100644
--- a/pkg/cgroups/manager/manager_test.go
+++ b/pkg/cgroups/manager/manager_test.go
@@ -169,7 +169,7 @@ func TestGetPodMetadataOnPodAdd(t *testing.T) {
 
        for _, tc := range tests {
                t.Run(tc.input.Name, func(t *testing.T) {
-                       mm.OnAddPod(tc.input)
+                       <-mm.onAddPod(tc.input)
 
                        got := mm.GetPodMetadataForContainer(tc.cgrpId)
                        require.Equal(t, tc.want, got)
@@ -203,13 +203,13 @@ func TestGetPodMetadataOnPodUpdate(t *testing.T) {
        require.Nil(t, got)
 
        // Add pod, and check for pod metadata for their containers.
-       mm.OnAddPod(pod3)
+       <-mm.onAddPod(pod3)
 
        got = mm.GetPodMetadataForContainer(c3CId)
        require.Equal(t, &PodMetadata{Name: pod3.Name, Namespace: pod3.Namespace, IPs: pod3Ipstrs}, got)
 
        // Update pod, and check for pod metadata for their containers.
-       mm.OnUpdatePod(pod1, newPod)
+       <-mm.onUpdatePod(pod1, newPod)
 
        got1 := mm.GetPodMetadataForContainer(c3CId)
        got2 := mm.GetPodMetadataForContainer(c1CId)
@@ -217,9 +217,7 @@ func TestGetPodMetadataOnPodUpdate(t *testing.T) {
        require.Equal(t, &PodMetadata{Name: pod3.Name, Namespace: pod3.Namespace, IPs: pod3Ipstrs}, got2)
 
        // Delete pod to assert no metadata is found.
-       mm.OnDeletePod(pod3)
-       // Wait for delete to be processed.
-       mm.wg.Wait()
+       <-mm.onDeletePod(pod3)
 
        got = mm.GetPodMetadataForContainer(c3CId)
        require.Nil(t, got)
@@ -231,14 +229,14 @@ func TestGetPodMetadataOnManagerDisabled(t *testing.T) {
        mm := newCgroupManagerTest(providerMock{}, cgroupMock{})
        c1CId := uint64(1234)
 
-       mm.OnAddPod(pod1)
+       <-mm.onAddPod(pod1)
 
        got := mm.GetPodMetadataForContainer(c1CId)
        require.Nil(t, got)
 
        // Enable the feature flag, but the cgroup base path validation fails.
        option.Config.EnableSocketLBTracing = true
-       mm.OnAddPod(pod1)
+       <-mm.onAddPod(pod1)
 
        got = mm.GetPodMetadataForContainer(c1CId)
        require.Nil(t, got)
@@ -259,7 +257,7 @@ func BenchmarkGetPodMetadataForContainer(b *testing.B) {
        mm := newCgroupManagerTest(provMock, cgMock)
 
        // Add pod, and check for pod metadata for their containers.
-       mm.OnAddPod(pod3)
+       <-mm.onAddPod(pod3)
 
        b.ResetTimer()
        for i := 0; i < b.N; i++ {
```

Reply:
(of course we could further optimize the above example to remove the channel allocation in the regular path, but I guess it depends how far it's worth iterating on this).

Reply:
All right. I changed the waiting logic a bit. One thing I also didn't like about the wait group approach is the dispersed wg.{Add, Done} calls. 
I added an events status channel that only the unit tests currently process. In future, we could reuse this channel to print last x events or something like that. 
Please check the latest revision.